### PR TITLE
ENH: Support masks in groupby prod

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #37493 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jorisvandenbossche 

This will have conflicts with #48018, will resolve after this is merged. Gives us the opportunity to simplify tests too

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -221,13 +222,13 @@ def _get_cython_vals(self, values: np.ndarray) -> np.ndarray:
             values = ensure_float64(values)
 
         elif values.dtype.kind in ["i", "u"]:
-            if how in ["var", "prod", "mean"] or (
+            if how in ["var", "mean"] or (
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py:225]: This is a micro optimization, but when hard coded I think it's better to use tuples than lists - they are stored as constants rather than constructed at runtime.

File: pandas/core/groupby/ops.py [hunk 2]
```diff
@@ -631,8 +640,8 @@ def _call_cython_op(
             # need to have the result set to np.nan, which may require casting,
             # see GH#40767
             if is_integer_dtype(result.dtype) and not is_datetimelike:
-                # Neutral value for sum is 0, so don't fill empty groups with nan
-                cutoff = max(0 if self.how == "sum" else 1, min_count)
+                # if the op keeps the int dtypes, we have to use 0
+                cutoff = max(0 if self.how in ["sum", "prod"] else 1, min_count)
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py:644]: I wonder if we should just skip this block entirely when `self.uses_mask()` is True - the checks below are O(n). But that'd be for a separate PR, this looks good to me.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looks great

File: pandas/core/groupby/ops.py (hunk 2 above)
Comment (phofl) [reply pandas/core/groupby/ops.py:644]: The post processing needs refactoring anyway when I am through with all algos. This is already on my todo list

**Review by rhshadrach — APPROVED** [review]
lgtm

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (phofl) [reply pandas/core/groupby/ops.py:225]: Forgot to reply here. Would also do this after we are through. Some of them won't be necessary anymore I think

**Comment by phofl** [comment]
cc @mroeschke  gentle ping

**Review by mroeschke — APPROVED** [review]
LGTM. Should be okay to sneak this one into 1.5, but probably better that the other groupby-masked ops go into 1.6

**Comment by mroeschke** [comment]
Thanks @phofl

**Comment by phofl** [comment]
Works for me, thx.

i will update the other prs after the ci is green again, makes reviewing easier
